beacon-chain/db/filesystem/data_column.go

```diff
@@ -0,0 +1,966 @@
+	}
+
+	file, err := dcs.fs.Create(filePath)
+	if err != nil {
+		return errors.Wrap(err, "data column sidecars file path create")
+	}
+
+	defer func() {
+		closeErr := file.Close()
+
+		// Overwrite the existing error only if it is nil, since the close error is less important.
+		if closeErr != nil && err == nil {
+			err = closeErr
+		}
+	}()
+
+	// Encode the SSZ encoded data column sidecar size.
+	var encodedSszEncodedDataColumnSidecarSize [encodedSszEncodedDataColumnSidecarSizeSize]byte
+	binary.BigEndian.PutUint32(encodedSszEncodedDataColumnSidecarSize[:], uint32(sszEncodedDataColumnSidecarRefSize))
+
+	// Concatenate the version, the data column sidecar size, the data column indices and the SSZ encoded data column sidecar.
+	countToWrite := headerSize + len(sszEncodedDataColumnSidecars)
+	bytes := make([]byte, 0, countToWrite)
+	bytes = append(bytes, byte(version))
+	bytes = append(bytes, encodedSszEncodedDataColumnSidecarSize[:]...)
+	bytes = append(bytes, indices[:]...)
+	bytes = append(bytes, sszEncodedDataColumnSidecars...)
+
+	countWritten, err := file.Write(bytes)
+	if err != nil {
+		return errors.Wrap(err, "write")
+	}
+	if countWritten != countToWrite {
+		return ErrWrongBytesWritten
+	}
+
+	return nil
+}
+
+// metadata runs file sanity checks and retrieves metadata of the file.
+// The file descriptor is left at the beginning of the first SSZ encoded data column sidecar.
+func (dcs *DataColumnStorage) metadata(file afero.File) (*metadata, error) {
+	// Seek to the beginning of the file.
+	_, err := file.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, errors.Wrap(err, "seek")
+	}
+
+	// Read the encoded file version.
+	var encodedFileVersion [versionSize]byte
+	countRead, err := file.Read(encodedFileVersion[:])
+	if err != nil {
+		return nil, errors.Wrap(err, "read file version")
+	}
+
+	if countRead != versionSize {
+		return nil, ErrWrongBytesVersionRead
+	}
+
+	// Convert the version to an int.
+	fileVersion := int(encodedFileVersion[0])
+
+	// Check if the version is the expected one.
+	if fileVersion != version {
+		return nil, ErrWrongVersion
+	}
+
+	// Read the SSZ encoded data column sidecar size.
+	var encodedSszEncodedDataColumnSidecarSize [encodedSszEncodedDataColumnSidecarSizeSize]byte
+	countRead, err = file.Read(encodedSszEncodedDataColumnSidecarSize[:])
+	if err != nil {
+		return nil, errors.Wrap(err, "read SSZ encoded data column sidecar size")
+	}
+	if countRead != encodedSszEncodedDataColumnSidecarSizeSize {
+		return nil, ErrWrongBytesDataColumnSidecarSizeRead
+	}
+
+	// Convert the SSZ encoded data column sidecar size to an int.
+	sszEncodedDataColumnSidecarSize := binary.BigEndian.Uint32(encodedSszEncodedDataColumnSidecarSize[:])
+
+	// Read the data column indices.
+	var indices [mandatoryNumberOfColumns]byte
+	countRead, err = file.Read(indices[:])
+	if err != nil {
+		return nil, errors.Wrap(err, "read data column indices")
+	}
+	if countRead != mandatoryNumberOfColumns {
```

Comment:
In all of these places where we allocate a certain size buffer, read it into it, and then check the bytes read, I think we could make things more concise using https://pkg.go.dev/io#ReadFull

Or even better, make a single buffer that is the right size to hold all the metadata and read it all at once, and just have a single error that indicates that the metadata section was truncated. Then you can slice and decode out of that single buffer. This avoids multiple syscalls.

Reply:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/aa84fbc597af4022e25d9ea104dca80c728849f2.

There is still `1+n` `file.Read`s:
- 1 (unique) for metadata, then
- 1 per column.

I did not find a way to, with only one syscall, read some parts of the file non contiguously.
We could of course read all the file at once then only select columns we want, but it would be a waste of read.
Also, we could do only one read for contiguous columns. (For example, if columns 4, 5, 6, 9 are requested, 1 read for columns 4 to 6 and one read for column 9. Let's say it's an optimisation for later ...)

Reply:
Fixed in https://github.com/prysmaticlabs/prysm/pull/15061/commits/22b3327a97afc4b1a171f569ce24520520b66a26.

Reply:
I only meant that we should get the metadata read done in one shot, not that we should read the column(s) data too. What you did is great!